lib/internal/streams/web.js

```diff
@@ -0,0 +1,4643 @@
+    return readRequest.promise;
+  }
+
+  releaseLock() {
+    if (!isReadableStreamDefaultReader(this))
+      throw new ERR_INVALID_THIS('ReadableStreamDefaultReader');
+    if (this[kState].stream === undefined)
+      return;
+    if (this[kState].readRequests.length) {
+      throw new ERR_INVALID_STATE.TypeError(
+        'Cannot release with pending read requests');
+    }
+    readableStreamReaderGenericRelease(this);
+  }
+
+  /**
+   * @readonly
+   * @type {Promise<void>}
+   */
+  get closed() {
+    if (!isReadableStreamDefaultReader(this))
+      return PromiseReject(new ERR_INVALID_THIS('ReadableStreamDefaultReader'));
+    return this[kState].close.promise;
+  }
+
+  /**
+   * @param {any} reason
+   * @returns {Promise<void>}
+   */
+  cancel(reason) {
+    if (!isReadableStreamDefaultReader(this))
+      return PromiseReject(new ERR_INVALID_THIS('ReadableStreamDefaultReader'));
+    if (this[kState].stream === undefined) {
+      return PromiseReject(new ERR_INVALID_STATE.TypeError(
+        'The reader is not attached to a stream'));
+    }
+    return readableStreamReaderGenericCancel(this, reason);
+  }
+
+  [kInspect](depth, options) {
+    return customInspect(depth, options, 'ReadableStreamDefaultReader', {
+      stream: this[kState].stream,
+      readRequests: this[kState].readRequests.length,
+      close: this[kState].close.promise,
+    });
+  }
+}
+
+class ReadableStreamBYOBReader {
+  [kType] = 'ReadableStreamBYOBReader';
+  /**
+   * @param {ReadableStream} stream
+   */
+  constructor(stream) {
+    if (!isReadableStream(stream))
+      throw new ERR_INVALID_ARG_TYPE('stream', 'ReadableStream', stream);
+    this[kState] = {
+      stream: undefined,
+      requestIntoRequests: [],
+      close: {
+        promise: undefined,
+        resolve: undefined,
+        reject: undefined,
+      },
+    };
+    setupReadableStreamBYOBReader(this, stream);
+  }
+
+  /**
+   * @param {ArrayBufferView} view
+   * @returns {Promise<{
+   *   view : ArrayBufferView,
+   *   done : boolean,
+   * }>}
+   */
+  read(view) {
+    if (!isReadableStreamBYOBReader(this))
+      return PromiseReject(new ERR_INVALID_THIS('ReadableStreamBYOBReader'));
+    if (!isArrayBufferView(view)) {
+      return PromiseReject(
+        new ERR_INVALID_ARG_TYPE(
+          'view',
+          [
+            'Buffer',
+            'TypedArray',
+            'DataView',
+          ],
+          view));
+    }
+    const viewByteLength = ArrayBufferViewGetByteLength(view);
+    const viewBuffer = ArrayBufferViewGetBuffer(view);
+    const viewBufferByteLength = ArrayBufferGetByteLength(viewBuffer);
+
+    if (viewByteLength === 0 || !viewBufferByteLength === 0) {
+      return PromiseReject(
+        new ERR_INVALID_STATE.TypeError(
+          'View ArrayBuffer is zero-length or detached'));
+    }
+    // Supposed to assert here that the view's buffer is not
+    // detached, but there's no API available to use to check that.
```

Comment:
Now that I think about it, if `viewBufferByteLength !== 0` then the view's buffer cannot be detached. So we don't actually need this assert in the specification. 😛